### PR TITLE
More selection streamlining

### PR DESCRIPTION
**Pull request by bstoeger**
### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [ ] New feature
- [x] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

This removes the peculiar way in which the map would call the dive list instead of the core function to select dives. For me, this makes selecting all visible dive sites of a huge dive-log usable. It's still dog-slow, but at least it doesn't hang from more than a minute!

This needs some serious testing!

### Changes made:

1) An assortment of code cleanups concerning the selection code.

### Related issues:

Hopefully improves #2776. I don't want to say fixes - because it is still dog-slow, but for me it is a distinct improvement.

### Mentions:

@janmulder, @atdotde, @dirkhh: this needs some serious testing - This part of the code is hairy and I want to add that it's only half our fault. The Qt selection code is ominous.

### CODE REVIEW

**Review by dirkhh — APPROVED** [review]
In my experience the only way this will get testing is by merging it.
I did a local build and can't say that I found anything that this breaks, but - meh... it's not something I ever use...
I'm planning to merge this all locally so we have just one push to the CI machinery

File: desktop-widgets/divelistview.cpp [hunk 1]
```diff
@@ -306,73 +306,6 @@ QList<dive_trip_t *> DiveListView::selectedTrips()
 	return ret;
 }
 
-void DiveListView::selectDive(int i)
```
Comment (neolit123) [inline desktop-widgets/divelistview.cpp:309]: this code goes back to after the gtk -> qt port and the addition of the marble map widget.

File: desktop-widgets/mapwidget.cpp [hunk 2]
```diff
@@ -94,7 +93,15 @@ void MapWidget::selectionChanged()
 void MapWidget::selectedDivesChanged(const QList<int> &list)
 {
 	CHECK_IS_READY_RETURN_VOID();
-	MainWindow::instance()->diveList->selectDives(list);
```
Comment (neolit123) [inline desktop-widgets/mapwidget.cpp:97]: this was the result of the marble -> qml map port.
at some point we had a compiler macro to smooth switch between the two.

File: desktop-widgets/divelistview.cpp (hunk 1 above)
Comment (neolit123) [reply desktop-widgets/divelistview.cpp:309]: is this and the plural not uses elsewhere outside of map logic?

File: desktop-widgets/mapwidget.cpp [hunk 3]
```diff
@@ -94,7 +93,15 @@ void MapWidget::selectionChanged()
...
+	// We get a list of dive indices, but the selection code wants a list of dives.
+	// Therefore, transform them here.
+	std::vector<dive *> selection;
+	selection.reserve(list.size());
+	for (int idx: list) {
+		if (dive *d = get_dive(idx))
+			selection.push_back(d);
+	}
+	setSelection(selection, current_dive);
```
Comment (neolit123) [inline desktop-widgets/mapwidget.cpp:104]: this is the first time i see:
https://github.com/Subsurface-divelog/subsurface/blob/7387b10b2944c0126999a3911d6ecf705d243a8b/core/selection.cpp#L151

if this now the way to change currentDive, LGTM.

Comment (dirkhh) [reply desktop-widgets/mapwidget.cpp:104]: so you're good with merging this, @neolit123 ?

Comment (neolit123) [reply desktop-widgets/mapwidget.cpp:104]: yes

**Comment by dirkhh** [comment]
manually merged

**Pull request closed without merging**